lib/src/ui/pages/profile/auth/auth_text_form_field.dart

```diff
@@ -0,0 +1,56 @@
+import 'package:allthenews/src/di/injector.dart';
+import 'package:allthenews/src/ui/pages/profile/auth/authorization_notifier.dart';
+import 'package:flutter/material.dart';
+
+abstract class _Constants {
+  static const fontFamily = 'Chomsky';
+  static const widgetPadding = 15.0;
+  static const borderRadius = 5.0;
+  static const focusedBorderSideWidth = 3.0;
+}
+
+class AuthTextFormField extends StatelessWidget {
+  final AuthorizationNotifier _authorizationNotifier = inject<AuthorizationNotifier>();
+  final TextEditingController textController;
+  final String hint;
+  final bool obscureText;
+
+  AuthTextFormField({
+    this.textController,
+    this.hint,
+    this.obscureText = false,
+  });
+
+  @override
+  Widget build(BuildContext context) {
```

Comment:
Zastanawiam się czy podejście jest poprawne, bo skoro ChangeNotifier służy do zarządzania stanem to wydaje mi się, ze powinno być tak, że kiedy np. wpisuje się coś do pola password, to to triggeruje metodę `notifier.updatePassword(password)`, co zmienia wewnatrz notifiera stan (`state.password`), który potem z powrotem propagowany jest na UI do TextFormField i analogicznie z walidacją, tj. po wciśnięciu przycisku wykonywana jest walidacja w notifierze, który w przypadku pustych pól wrzuca do viewState błąd `Pole nie może być puste`, który dodawany jest do TextFormFielda

Reply:
zauważ, że użyłem tu podejścia z widgetem `Form` - https://flutter.dev/docs/cookbook/forms/validation
to on przechowuje stan walidacji pod spodem i używa to tego klucza 
```
final GlobalKey<FormState> _formKey = GlobalKey<FormState>();
```
notifier posiada tylko metodę to walidacji pól, ale postanowiłem lokalnie walidować tylko niepustość i dlatego każde pole ma taką samą metodę walidującą. Podejście o którym mówisz sprawdziłoby się gdyby przechowywanie stanu walidacji było całkowicie po naszej stronie

Reply:
znaczy ja rozumiem to podejście, tylko chodzi mi o to, że jeżeli mamy architekturę, w której notifier propaguje stan na ui to powinien propagować wszystko, bo teraz część stanu jest zarządzana przez notifiera, a część przez sam ui :thinking: Dokumentacja flutterowa nie uwzględnia żadnej architektury. W necie za bardzo nie ma użycia providera z formularzem i jego walidacją (znalazłem jeden ale gość pomieszał bloc z providerem :P), ale chodzi mi o mniej więcej coś takiego jak w samplach flutter_bloc:
https://github.com/felangel/bloc/blob/master/examples/flutter_form_validation/lib/main.dart

Reply:
w tym przykładzie co pokazałeś w ogóle nie jest używany widget `Form`, który ogarnia stany `TextFormFieldów` OOTB. Tak naprawdę jedyny powód, dla którego są tam `TextFormFieldy` to atrybut `initialValue` (który można by było też ograć na zwykłym `TextFieldzie`). Moim zdaniem jeżeli jest jakaś funkcjonalność dostępna przez framework to warto z niej skorzystać, jeżeli chcielibyśmy wszystko trzymać w stanie to równie dobrze pozycje scrolla moglibyśmy tam przechowywać, tylko po co? 

Ja sugerowałem się przykładem z exampli do flutter_auth -> https://github.com/FirebaseExtended/flutterfire/blob/799231ac57a3005a4fbc8fd9c5951c3943e4a286/packages/firebase_auth/firebase_auth/example/lib/register_page.dart

Reply:
Ogólnie stan we flutterze można cały trzymać na ui i są wszystkie niezbędne narzędzia do tego, żeby zarządzać stanem tak jak to zostało zaprojektowane. Nie wymusza żadnej architektury. Jako, że nie wymusza architektury to powstało x różnych, które służą zarządzaniem stanu. Ten przykład z flutterfire nie ma żadnej architektury. Stan można ograć na UI, ale mimo wszystko stworzyłeś HomeNotifier (i słusznie), który wyciąga stan do notifiera, gdzie setState też to ogarnia, ale z notifierem jest zdecydowanie czyściej. Jeżeli przyjęta architektura wyciąga stan do notifierów i propaguje ten stan na ui to powinno być to konsekwetne. Jeżeli trzeba by było zarządzać samemu stanem scrolla to imo również powinien być propagowany z notifiera, ale na szczęście nie trzeba. Natomiast stan pól powinien być przechowywany w notifierze, ponieważ jest częścią stanu UI, który musi być zarządzany - tj. walidowany i wysyłany na backend. Sama walidacja też powinna odbywać się w walidatorze, który byłby gdzieś w domenie i notifier by sobie nim walidował pola przy kliknięciu buttona, a następnie wrzucił do viewState informację czy ma wyświetlić błedy walidacyjne i jak tak to które. W tym podejściu nie ma znaczenia czy walidacja jest skomplikowana czy nie, bo nawet jakby się skomplikowała to dla widgetów jest to niezauważalne, przeływ danych jest identyczny :) Kolejna kwestia to, że trzymanie wszystkiego w jednym miejscu (notifierze) pozwoli przetestować logikę klasę w izolacji od UI :)

Reply:
trochę nie widzę tej granicy kiedy korzystać z narzędzi ogarniających stan OOTB, a kiedy już samemu tym zarządzać. W tym przypadku widgety `Form` i `TextFormField` stają się bezużyteczne. Jeżeli chcemy mieć stan w notifierze + korzystać z tego co oferuje framework to może trzymać w stanie notifiera sam `GlobalKey` ?

Reply:
Jeżeli jedyną wartością Form i TextFormField jest zarządzanie stanem na ui to przy obecnej architekturze rzeczywiście są bezużyteczne, ale nie powiedziałbym, że to coś złego. Wiesz, w androidzie jak robisz MVP to nie pchasz LiveDaty na UI tylko dlatego, że jest coś takiego, tylko trzymasz się założeń. Różnica między tym co mamy a blocem jest taka, że tam wszystko opiera się na strumieniach i powiedzmy, że jest większe wyróżnienie istoty eventów propagowanych z ui do bloca. I wg sampli też nie korzysta z walidacji, którą daje Form i to też nic złego. Takie podejście fajnie się sprawdza na pewno przy `Simple app state management` (na zakładce `List of state management approaches` na flutter.dev), ale założenia, które przyjęliśmy przy providerze są jednak inne. Imo nie ma co mieszać dwóch podejść i ładnie zrobić stan samemu, bo wpychanie GlobalState, który rozszerza `State<StatefulWidget>>` to trochę jednak wpychanie UI wyżej (do notifiera) niż powinien być (UI)

Reply:
ok, przekonałeś mnie, zrobię zarządzanie stanem walidacji z poziomu notifiera 😄